### ISSUE TITLE
Provide a sample integrating ASP.NET Identity

### ISSUE THREAD

**Issue by AlexZeitler** [aspnet-contrib/AspNet.Security.OpenIdConnect.Server#94]
Do you plan to provide a sample showing usage with local users (in memory would be good enough :smile:) in addition to twitter and fb?

**Comment by PinpointTownes** [comment]
Yep, that would be a nice idea. I guess it would be implemented using ASP.NET Identity.

In the meantime, you can also have a look at the Nancy sample (for OWIN/Katana 3), that uses a custom authentication module offering a good/old username/password authentication: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/tree/dev/samples/Nancy :smile:

**Comment by AlexZeitler** [comment]
Thanks, I'll have a look at it and try to get it working.

**Comment by PinpointTownes** [comment]
Note that it's really trivial: it doesn't check anything and has no database to store users' credentials. I didn't want to focus the sample on local authentication :smile:

**Comment by AlexZeitler** [comment]
No problem. My main goal is to a OpenId Connect server running on ASP.NET vNext for an ASP.NET vNext API with an Angular client. All hosted on a Linux box. :smile:

**Comment by PinpointTownes** [comment]
Cool :sunglasses: 

Creating an AngularJS sample is also on my todo list, but I must admit I'm a total JS noob :smile: 
If you want to contribute by sending a PR for that, don't hesitate :tada:

**Comment by AlexZeitler** [comment]
Ok I'll focus on it as it is a requirement of the current long term project. The AngularJS part is the trivial one for me here :)

**Comment by AlexZeitler** [comment]
How does the Nancy Sample (I've never used NancyFx) validate the credentials? I've found the `SignIn.cshtml` but where does the magic happen?

**Comment by AlexZeitler** [comment]
Found it... @ `AuthenticationModule.cs`

**Comment by PinpointTownes** [comment]
Actually, it just checks you correctly entered a username: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/dev/samples/Nancy/Nancy.Server/Modules/AuthenticationModule.cs#L17

**Comment by AlexZeitler** [comment]
Dump question: Whats the corrent `AuthenticationScheme` here (for local users in the `SignIn` method of the `AuthenticationController`)? `Bearer`? `Basic`?
````
Context.Authentication.SignIn(<AuthenticationScheme>, principal, properties);
`````

`NancyContext` and MVC `HttpContext` are different...

**Comment by PinpointTownes** [comment]
`ServerCookie`: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/dev/samples/Nancy/Nancy.Server/Modules/AuthenticationModule.cs#L34

You can change it in the cookies authentication middleware options: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/dev/samples/Nancy/Nancy.Server/Startup.cs#L37

**Comment by AlexZeitler** [comment]
Ok. This works but the redirect to the redirect url does not happen without any errors. If I send the client app to the sign in again, it asks me if I want to grant the client access and if I allow, it gets redirected back and has the id_token. Logging out works as well. 

What I see is that the request to `/connect/authorize` is redirected to `Location: http://localhost:54540/signin?ReturnUrl=%2Fconnect%2Fauthorize%3Funique_id%3DYndZd3NEUGVOb2ZHVU1oQ3F1UzlUVklBOU1XV3AzOGdHckNxMnlFdjZCST0` so the question is where can I change the `ReturnUrl` being used here. 

Or am I doing it completele wrong? :astonished: 

The source of the `SignIn` method of the `AuthenticationController` looks like this:

````
 if (!string.IsNullOrWhiteSpace(model.Username) && model.Username == model.Password)
            {
                var properties = new AuthenticationProperties
                {
                    RedirectUri = model.ReturnUrl
                };

                var identity = new ClaimsIdentity("ServerCookie");
                identity.AddClaim(new Claim(ClaimTypes.Name, model.Username));
                identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, model.Username));

                var principal = new ClaimsPrincipal(identity);

                Context.Authentication.SignIn("ServerCookie", principal, properties);
            }

                return new ChallengeResult(model.Provider, new AuthenticationProperties {
                RedirectUri = model.ReturnUrl
            });
````

**Comment by PinpointTownes** [comment]
Is your app hosted somewhere?

> What I see is that the request to /connect/authorize is redirected to Location: http://localhost:54540/signin?ReturnUrl=%2Fconnect%2Fauthorize%3Funique_id%3DYndZd3NEUGVOb2ZHVU1oQ3F1UzlUVklBOU1XV3AzOGdHckNxMnlFdjZCST0 so the question is where can I change the ReturnUrl being used here.

Being redirected to the `SignIn` endpoint is not surprising if you're not authenticated yet. Do you mean that you're seeing 302 responses even when you're connected?

Unlike most apps where the `ReturnUrl` parameter is automatically managed by the cookies middleware (which is also responsible of redirecting you when you hit a protected page without being authenticated), it is manually computed in the authorization controller: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/vNext/samples/Mvc/Mvc.Server/Controllers/AuthorizationController.cs#L55.

**Comment by AlexZeitler** [comment]
After signing in, I'm redirected to this URI:
`http://localhost:54540/signin?ReturnUrl=%2Fconnect%2Fauthorize%3Funique_id%3DYndZd3NEUGVOb2ZHVU1oQ3F1UzlUVklBOU1XV3AzOGdHckNxMnlFdjZCST0`
Which results in a 400:

`````
HTTP/1.1 400 Bad Request
Cache-Control: no-cache
Pragma: no-cache
Expires: -1
Server: Microsoft-IIS/10.0
Content-Security-Policy: default-src 'self'; script-src 'unsafe-inline'; style-src 'self' 'unsafe-inline'; img-src 'self' *
X-Content-Type-Options: nosniff
X-Frame-Options: Deny
X-XSS-Protection: 1; mode=block
Set-Cookie: .AspNet.ServerCookie=CfDJ8JPaoFi5VXZHgRH00dte13BTeYTwkWSgPAjCtks-MvllwcBFve_YmGjmMeYoAA4YppBl0ZNb21zvqBZraxqJQINUJHnxwCf-qSrkETkdhBpCvlUIJMHiB7rkwm9IhONDTT44l0nypzr8NeSVnGDsnP8xaTEF-o8m1A11egOAxWRlMC7-FRLb1USc6xM3s0_nQyX7Qn5kgJgyZt5EDit1QTG9qsGZ8IKyr0cmzgrgIigKC97EhAUVVmcBC_xSp1sLGwdVwOn2CpJgZpzxw3Sko047yjqCKohg-0mzXVZqSmBzncgjVdFwfiGLH8XhRhawQ0590mDFiED46ZPY7eSW6nHftzMx_gvsxIIm_6Ll1Mt_mzkg4Xogz7xkhRThEJ9Eyh3w-tc8Ch5b3cmgvGcexI1slJVn8ZPPbyCmvzGOZDKxXRCz59ptE9wosFYgrntIbldTSQmshJDUP1By-w25fCuOSbylNne01gS0q8PbPOAbE03tyfw4b05w24VWjh1_2Rq-ljxGme4BVgmyupwf_MA; path=/; httponly
X-SourceFiles: =?UTF-8?B?QzpcVXNlcnNcQWxleGFuZGVyIFplaXRsZXJcc3JjXGFzcG5ldDVvcGVuaWRcQXNwTmV0LlNlY3VyaXR5Lk9wZW5JZENvbm5lY3QuU2VydmVyXHNhbXBsZXNcTXZjXE12Yy5TZXJ2ZXJcd3d3cm9vdFxzaWduaW4=?=
X-Powered-By: ASP.NET
Date: Sun, 28 Jun 2015 18:25:35 GMT
Content-Length: 0
````
I would expect to be redirected to the consent page.

**Comment by PinpointTownes** [comment]
Could you please upload a repro? It would be easier to help you :smile:

**Comment by AlexZeitler** [comment]
Sure, here we go: https://github.com/AlexZeitler/AspNet.Security.OpenIdConnect.Server

The JavaScript client is this one:
http://brockallen.com/2015/06/19/demos-ndc-oslo-2015/

Just follow the first OneDrive-Link and download the sample, then use the `1_id_tokens` project and modify the following lines in `index.html`:

(81) `var client_id = 'myClient';`

(84) `var authorizationUrl = 'http://localhost:54540/connect/authorize';`

(219)
````
 function logout() {
            //window.location = "http://localhost:54540/connect/logout";
            window.location = "http://localhost:54540/connect/logout?post_logout_redirect_uri=" + encodeURIComponent(window.location.protocol + "//" + window.location.host + "/index.html") + "&id_token_hint=" + encodeURIComponent(id_token);
        }
````

**Comment by AlexZeitler** [comment]
Just tested the JavaScript client with the Nancy Example and the flow works fine here.

**Comment by PinpointTownes** [comment]
Your customized `SignIn` has a 'little' bug.

You forgot to stop processing the request after the `Context.Authentication.SignIn` call: the rest of the method is invoked, `string.IsNullOrWhiteSpace(model.Provider)` returns `true` (because you selected local authentication), and it returns a 400 response :smile: 

Simply skip the default logic when selecting "local authentication" and it should work:

![image](https://cloud.githubusercontent.com/assets/6998306/8398056/fc2621da-1de1-11e5-9f3c-c06b964b9b3f.png)

**Comment by AlexZeitler** [comment]
Stuped me :frowning: 
Thanks (once again) for helping :+1: 

Yet, another few questions: 

1. Currently the consent screen appears every time I log in - can I show it only once (the first time)?
2. Can I disable the consent screen completely?
3. Is it possible to run the resource controller on a different host?

**Comment by AlexZeitler** [comment]
Another one :smile: 
You mentoined documentation in another issue.
Where can I find it - or did you mean the code comments?

**Comment by PinpointTownes** [comment]
> Stuped me :frowning: 
Thanks (once again) for helping :+1:

You're welcome! :smile: 

> Currently the consent screen appears every time I log in - can I show it only once (the first time)?

Unlike turnkey solutions (like IdentityServer for instance), the authorization consent part is entirely left as an exercise: your imagination should be the only limit here :smile: 

As you already figured out, the `Mvc.Server` sample doesn't store the authorizations granted by a specific user and always displays a consent form. Of course, this can be easily added by adding a new `Authorization` model, by modifying `Authorize` to determine whether a given user has previously given his consent or not, and by updating `Accept` to store the granted authorizations in the database.

> Can I disable the consent screen completely?

If you don't want to display a consent form, simply call `Context.Authentication.SignIn` in the `Authorize` action, and the user will be redirected to the client app. You can also do that directly in your authorization provider, in the `AuthorizationEndpoint` notification:

```csharp
public override Task AuthorizationEndpoint(AuthorizationEndpointNotification notification) {
    var principal = new ClaimsPrincipal();

    // The user will be immediately redirected to the client application
    // without being displayed a consent form.
    notification.HttpContext.Authentication.SignIn("Bearer", principal);
    notification.RequestCompleted();

    return Task.FromResult<object>(null);
}
```

You should really be careful about this kind of feature, as it definitely represents a security risk, specially when using the implicit or hybrid flows (I know that skipping the consent form was/is massively used by the Thinktecture guys, but I'm not sure they realize all the implications it has). 

You can mitigate that using `id_token_hint`: flow the `id_token` you received during a previous authentication flow to the authorization endpoint and ensure it corresponds to the user making the authorization request:

```csharp
[HttpGet("~/connect/authorize")]
[HttpPost("~/connect/authorize")]
public async Task<IActionResult> Authorize(CancellationToken cancellationToken) {
    // Extract the identity corresponding to the identity token retrieved from the id_token_hint parameter.
    var principal = await Context.Authentication.AuthenticateAsync("Bearer");
    if (principal != null) {
        var identifier = principal.Principal.FindFirst(ClaimTypes.NameIdentifier);
        if (string.Equals(identifier?.Value, User.FindFirst(ClaimTypes.NameIdentifier)?.Value, StringComparison.Ordinal)) {
            // Only skip authorization consent when the user corresponds to the one represented by the id_token.
        }
    }
}
```

> Is it possible to run the resource controller on a different host?

Yep. It will even work natively if you use the OAuth2 bearer middleware, that automatically downloads the OIDC configuration it needs to validate access tokens:

```csharp
app.UseOAuthBearerAuthentication(options => {
    options.AutomaticAuthentication = true;
    options.Audience = "[the URL of your resource server]";
    options.Authority = "[the URL of your authorization server]";
});
```

> You mentoined documentation in another issue (don't want to bother you constantly with basic questions).
Where can I find it - or did you mean the code comments?

Sadly, there's no documentation yet (except code comments). It's definitely on my todo list, but I lack free time for that :cry:

**Comment by AlexZeitler** [comment]
Thanks, will dig through your answers tomorrow. Maybe I can help with the documentation after I'm done with the AngularJS sample. That would make sure I got everything right and you get the documentation.

**Comment by PinpointTownes** [comment]
> Maybe I can help with the documentation after I'm done with the AngularJS sample. That would make sure I got everything right and you get the documentation.

That would be great, thanks! :smile: 

Out of curiosity, would you be interested in a companion project offering a "turnkey approach" like IdSrv? I've always considered that using IdSrv was the way to go when you wanted something "ready to use" without writing custom code, but maybe it's time to change my mind :smile:

**Comment by AlexZeitler** [comment]
Well, I've been using IdentityServer3 in several projects when running on the Windows stack.
I also tried to get it running on Linux, but it didn't work (might be my fault).
From what I've read in the issues on GH, the strategy seems to be to move it to vNext when vNext is done or at least more stable (which is totally viable as everyone has to get it's business done first).
As already mentioned, for the current project I definitely need a solution that runs on Linux today (even it is not complete nor 100 percent stable) because I need to have it run in Docker containers.
Interestingly there seems to be no OSS OpenId Connect Server in the NodeJS world (at least I didn't find one after digging through Google for a week) and thats what might make your turnkey solution even more interesting when being run in a container for those guys.
Regarding IdentityServer I think that having two options is better than one and I've talked to people that will not use IdentityServer for some reasons but also look for alternative solutions.

**Comment by AlexZeitler** [comment]
I've been working on the Angular demo client and noticed that, when using implicit flow as configured in the default `Mvc.Server` sample, `expires_in` is not send back to the client as you can see from `this query fragment:

`nonce=feb1277c73b51903511c5e05696ae8ca&state=5adf45fcf96e6d02236dd534b08418b0&token_type=bearer&access_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjA1MkM1RTQyMTVDRDBDMUUxNTA1RTA4RTZBRjNBREJFRkJGRDc4MjIifQ.eyJuYW1laWQiOiIxNjI4NzE1MiIsInVuaXF1ZV9uYW1lIjoiYWxleHplaXRsZXJfIiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo1NDU0MC8iLCJleHAiOjE0MzU4NzY1MjQsIm5iZiI6MTQzNTg3MjkyNCwiYWN0b3J0IjoiZXlKMGVYQWlPaUpLVjFRaUxDSmhiR2NpT2lKdWIyNWxJbjAuZXlKdVlXMWxhV1FpT2lKdGVVTnNhV1Z1ZENJc0luVnVhWEYxWlY5dVlXMWxJam9pVFhrZ1kyeHBaVzUwSUdGd2NHeHBZMkYwYVc5dUluMC4ifQ.NbkKPXxQpblq0AzbKw3c8EiOuVsHGSxbwXz5I6CqV35UN42wa-ZHC6UGOQYz7jfjYFUOWSES3AiBwr2dk4cUJKqrLPPICxwqW_K0LAsvNhXFpmJaeGnM6YrJt8d-ygQCtGiLMMF6aqzuuAqeIEbcLWvx3fBnUg2P4BCxxzkoj_VHIvC8Jr1gYNqO_C4gxAUHF-1XiR2kG-6A8d61FNCjqQEU_2cGC1JgHDx97P0WRGJ8zyHIKemj018WPQAvrtu6mGsTlxDcwTOn5bw-9aB4HFx1UAmRK4bc-5JGevndxMu5pXZXpSj_zuuYg26hY9uU-6cP6An2rlfleWZ8sn_ARg&id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjA1MkM1RTQyMTVDRDBDMUUxNTA1RTA4RTZBRjNBREJFRkJGRDc4MjIifQ.eyJuYW1laWQiOiIxNjI4NzE1MiIsInVuaXF1ZV9uYW1lIjoiYWxleHplaXRsZXJfIiwiaWF0IjoiMTQzNTg3MjkyNCIsImF0X2hhc2giOiI5NnRqdlpVMlNQaTZHMUF6bnJYekx3Iiwibm9uY2UiOiJmZWIxMjc3YzczYjUxOTAzNTExYzVlMDU2OTZhZThjYSIsInN1YiI6IjE2Mjg3MTUyIiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo1NDU0MC8iLCJhdWQiOiJteUNsaWVudCIsImV4cCI6MTQzNTg3NDEyNCwibmJmIjoxNDM1ODcyOTI0LCJhY3RvcnQiOiJleUowZVhBaU9pSktWMVFpTENKaGJHY2lPaUp1YjI1bEluMC5leUp1WVcxbGFXUWlPaUp0ZVVOc2FXVnVkQ0lzSW5WdWFYRjFaVjl1WVcxbElqb2lUWGtnWTJ4cFpXNTBJR0Z3Y0d4cFkyRjBhVzl1SW4wLiJ9.qUlqH5SQn4b02UOG94wHsh67NZubdPnKrNz-63S2JqnxFFYGjFEQGYbenkje1vObT_WNjf716uTrKvKsKfLeR-HJKKyy_8_dRprYtzUEU_DvjZazCKYY3AxwLb1KSVZINClTfy-gHU3Ya3WEiAT-ogCkrTz7-ErZKAllvLEpzzY2v4NoS8MBTp9GfthHrV8idGbKdAMBw7dl7XN21k7JfUd_qnr2S-PfXuaqeWXWmzWW4LG9bJkr4Ru6WMZv2fyXG13jmkOIZCEQQAzdhuBKibswKVAHL7EZrJKToBHV17NGdd57sUZPB6eZAurYoTPdRyXtFZfuR4rGzvAFE5-6QQ`

Is there a way to activate the expiration for implicit flow?

**Comment by PinpointTownes** [comment]
Hum, sounds like a regression (though `expires_in` is purely optional for an authorization response).

I'll fix it. Thanks for pointing it out :smile:

**Comment by AlexZeitler** [comment]
That is my opinion, too. But the Angular OAuth component I'm using seems to treat it as a mandatory property.

**Comment by PinpointTownes** [comment]
That's a zealous library :smile: 

![image](https://cloud.githubusercontent.com/assets/6998306/8488873/985e6b70-2116-11e5-9912-a92de3f32ae7.png)

![image](https://cloud.githubusercontent.com/assets/6998306/8488878/b2445cde-2116-11e5-9b05-782eecb67e85.png)

https://tools.ietf.org/html/rfc6749#section-4.2.2
http://openid.net/specs/openid-connect-core-1_0.html#ImplicitAuthResponse

**Comment by AlexZeitler** [comment]
I think I'll use it only temporarily or I might send them a PR.

**Comment by csunali** [comment]
Hi there,

I am never able to get the LogoutEndpoint method executed. The server options the following line of code in it:

`options.TokenEndpointPath = new PathString("/Auth/Token")`

Even if I implement a controller that matches the above path string and the custom action gets executed (so that I can logout the user), the server provider's LogoutEndpoint method still does not get executed.

What do you think I might be missing or doing wrong?

**Comment by PinpointTownes** [comment]
Hey @csunali,

Sorry to hear that you're in trouble with the new logout endpoint.
Could you please open a new ticket and share your Startup.cs configuration and your logout controller?

Thanks!

**Comment by csunali** [comment]
Hey @PinpointTownes,

Just did that. Thank you in advance!

P.S.: Your code was of great help to implement bearer authentication.

**Comment by PinpointTownes** [comment]
@csunali thanks! :smile:

**Comment by AlexZeitler** [comment]
Hm.. got another issue:
When calling the `api/message` endpoint, it throws the following exception:

`IDX10214: Audience validation failed. Audiences: 'empty'. Did not match:  validationParameters.ValidAudience: 'http://localhost:54540/' or validationParameters.ValidAudiences: 'null'`

The token contains ` "iss": "http://localhost:54540/"`

**Comment by PinpointTownes** [comment]
@AlexZeitler that's by "design": if you don't explicitly specify it, the authorization server has no way to determine the destination/audience of an access token, which is issued without the `aud` claim required by default by the OAuth2 bearer middleware. You have 2 options to solve this issue:

* Specify the `resource` parameter when making the authorization request. This way, the authorization server knows exactly which audience (API) it should add in the `aud` claim: it must corresponds to the `Audience` property of the OAuth2 bearer middleware. You can specify multiple API by separating their address by a space in the `resource` parameter.

* Disable audience validation (`options.TokenValidationParameters.ValidateAudience = false`) in the OAuth2 bearer middleware options.

**Comment by AlexZeitler** [comment]
Thanks, `options.TokenValidationParameters.ValidateAudience = false` did fix it. Need to investigate this a bit further.

**Comment by AlexZeitler** [comment]
Just read this: http://openid.net/specs/openid-connect-implicit-1_0.html#IDToken

Regarding `aud` it says:

> REQUIRED. Audience(s) that this ID Token is intended for. It MUST contain the OAuth 2.0 client_id of the Relying Party as an audience value. It MAY also contain identifiers for other audiences. In the general case, the aud value is an array of case sensitive strings. In the common special case when there is one audience, the aud value MAY be a single case sensitive string.

So if I don't specify the `resource` parameter it should at least contain the value of the `client_id` param which has been send, right?

**Comment by PinpointTownes** [comment]
Don't mix up the identity token with the access token: :smile:

* The identity token (that MUST be a JWT token) is primarily used by the client application, that's why it always includes an `aud` containing the `client_id`. The `resource` parameter has absolutely no influence on the identity token.

* The access token (that MAY be a JWT token), is used by the resource server. If you don't specify an explicit `resource` parameter, we have no way to determine the API you'll use it for, and that's why it's issued without any audience (which means that the access token is virtually acceptable by every resource server accepting your tokens), which is a perfectly legal scenario:

![image](https://cloud.githubusercontent.com/assets/6998306/8502694/5ad6d410-21b4-11e5-830b-e46c8cc55707.png)

https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32#section-4.1.3

**Comment by AlexZeitler** [comment]
Ok.
IdentityServer does handle it this way (just tried it) - also no `resource` param:

`id_token`
`"aud": "sampleapiclient"` whereas `sampleapiclient` has been the `client_id` in the authorization request.

`access_token`
`"aud": "http://localhost:44309/identity/resources"` whereas `localhost:44309/identity`is the base URI of the IdP.

This is the full `Location` header of the `302` response after calling the authorization endpoint:

`http://localhost:9000/#id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6ImEzck1VZ01Gdjl0UGNsTGE2eUYzekFrZnF1RSIsImtpZCI6ImEzck1VZ01Gdjl0UGNsTGE2eUYzekFrZnF1RSJ9.eyJub25jZSI6ImUyNDYxZDlhYmZlMTFkZDZhMjAzODI3ZDc5YTg4OTdkIiwiaWF0IjoxNDM1OTQyNjYxLCJhdF9oYXNoIjoiSVE1NE13d1JVeXBoSEQybEdwZGVmUSIsInN1YiI6IjEiLCJhbXIiOiJwYXNzd29yZCIsImF1dGhfdGltZSI6MTQzNTk0MjY2MSwiaWRwIjoiaWRzcnYiLCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjQ0MzA5L2lkZW50aXR5IiwiYXVkIjoic2FtcGxlYXBpY2xpZW50IiwiZXhwIjoxNDM1OTQyOTYxLCJuYmYiOjE0MzU5NDI2NjF9.GRZ8VnPLPcB3j3VZ4l1lG8CR1fYfo-t4MWJSpuo0AqiDqMpHjQzVPeinQ_-5KY6E1-vM5WBxrTWuy3lvU9rN-ZLfq-Gb_7uSuOJHRa1tPInIN71_kaWA_9sAmtAxOzN0DBqmyfhOoKMdUtdFmhS7K0dDF57rd10DR5BAvT7t3CpHv6hZ4OkGmNacPr-8FNsIYNeNI_GYez60E_o04fZrK5Ro_MXshjNq0rr1klfIqQtvt-G8BDZJTK0xal7k4OCmOVZGRXqeDlXghj15TZGJI9PtXfBdu20VbtdrMtmuSEgF-2vXcqXaP0tYVvoofXYYciUxXQ-q5ludk44Zkr1KAg&access_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6ImEzck1VZ01Gdjl0UGNsTGE2eUYzekFrZnF1RSIsImtpZCI6ImEzck1VZ01Gdjl0UGNsTGE2eUYzekFrZnF1RSJ9.eyJjbGllbnRfaWQiOiJzYW1wbGVhcGljbGllbnQiLCJzY29wZSI6WyJvcGVuaWQiLCJteVdlYkFwaSJdLCJzdWIiOiIxIiwiYW1yIjoicGFzc3dvcmQiLCJhdXRoX3RpbWUiOjE0MzU5NDI2NjEsImlkcCI6Imlkc3J2IiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo0NDMwOS9pZGVudGl0eSIsImF1ZCI6Imh0dHA6Ly9sb2NhbGhvc3Q6NDQzMDkvaWRlbnRpdHkvcmVzb3VyY2VzIiwiZXhwIjoxNDM1OTQ2MjYxLCJuYmYiOjE0MzU5NDI2NjF9.bE1qGRhY0mdsi-BegvPyxzsn0aorKtGDt3REFha1Lo3OMOlyjso3oBIV9nkjkqYGnXUSzAs15DzpUwcByzaugZyHnD1hVxjcMhFkTrFwwRRsunyHde_WptkzSZqvS67u6TkV7pAc5U3xec-9-ho9o6oNFWlbmKQmW_iR3ToOqnxGbTAouOSr8IWxVXU1Cos1hxFZql8WkZli4PUNBybegSrgIMLbEQ2lhBKLmH2k7u5T-COqGGCIgAHfbBphYRYkF390c4MuyQ2c7aMMVvFBMF0Zf4BvBApyUEP06NpfbrxXkCK_CjGo_v9ESUTrZq5O4ZOyUuhFkzd1Ex4NTs5znA&token_type=Bearer&expires_in=3600&scope=openid%20myWebApi&state=4eb82a34dde1d734f2a99e6a30c35146&session_state=M1ZuUPp4wmL-sCoTVVoSVy1JUqpfdX90eFw012H4KlU.0672359759369c78aaf0a292096c13c0`

**Comment by PinpointTownes** [comment]
And there's an endpoint listening at `http://localhost:44309/identity/resources`?
(I don't remember, does IdSrv support `resource`?)

**Comment by AlexZeitler** [comment]
No, `http://localhost:44309/identity/resources` returns a `404`.

**Comment by PinpointTownes** [comment]
Then, I guess Dominick or Brock preferred adding a 'dummy' `aud` claim rather than issuing an access token without any audience :smile: 

This approach has clearly a downside: according to the JWT specs, a resource server would HAVE TO reject such an access token to be fully specs-compliant, even when disabling audience validation. I'm definitely not convinced it's the right approach.

**Comment by AlexZeitler** [comment]
This gets interesting :smile: as IdentityServer has been certified by OpenId: http://openid.net/certification/

**Comment by AlexZeitler** [comment]
Did you already try to run http://openid.net/certification/testing/ against AspNet.Security.OpenIdConnect.Server?

**Comment by PinpointTownes** [comment]
> This gets interesting :smile: as IdentityServer has been certified by OpenId: http://openid.net/certification/

Well, the conformance tests don't validate access tokens, as you're totally free to choose your own format (they can even be totally opaque).

And these tests don't cover the whole specification: AFAIK, IdSrv doesn't support POST authorization requests, though that's a mandatory feature:

![image](https://cloud.githubusercontent.com/assets/6998306/8503050/4403557e-21b9-11e5-87e1-fa3c9a6270a2.png)

> Did you already try to run http://openid.net/certification/testing/ against AspNet.Security.OpenIdConnect.Server?

Nope, not yet :smile: 

That said, `AspNet.Security.OpenIdConnect.Server` is a bit special, as some parts are delegated to the end user (like the consent form or the logout endpoint), who's free to implement them respecting the specifications or not. An implementation A might pass the tests, while another one might not, even if they used the same library.

I guess that would be a nice thing to make sure our `Mvc.Server` sample pass the conformance tests. Do you wanna try?

**Comment by AlexZeitler** [comment]
Sure. But first I need to test if the Angular client scenario works as is on Linux. I think I'll do this this evening.

If it works, I'll send you a PR.

BTW: the resource param has to be send formurlencoded with client_id et al?

**Comment by PinpointTownes** [comment]
> BTW: the resource param has to be send formurlencoded with client_id et al?

If you're using POST authorization requests, yes, it must be added to the body and must respect the `application/x-www-form-urlencoded` format.

For GET requests, it's also preferable to 'URL-encode' it.

**Comment by AlexZeitler** [comment]
Thanks. Is the resource param specific to OIDC or is this also part of the OAuth 2 spec? Just asking as the Angular component doesn't seem to support resource.

**Comment by PinpointTownes** [comment]
It's a totally non-standardized thing, so it's not surprising :smile:

Some people hate it (*cough* like Dominick *cough*), others love it.
If you need a reference, `resource` is massively used by Azure AD, and `AspNet.Security.OpenIdConnect.Server` reuses roughly the same approach.

**Comment by AlexZeitler** [comment]
Ok, this explains why I didn't find it here: http://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowAuth

**Comment by AlexZeitler** [comment]
But if it as non-standard thing, shouldn't the out-of-the-box solution provide an Audience value?
I'm a bit confused at the moment 😄

**Comment by PinpointTownes** [comment]
Which value would you like to use? :smile:

**Comment by AlexZeitler** [comment]
I'm not sure at the moment. I need to think about it. Issue 279 at IdentityServer also has a discussion related.

**Comment by AlexZeitler** [comment]
Just for understanding: JWT validation requires you to pass a resource param. 
OpenId Connect allows you to use JWT but does not enforce resource param to be able to create valid access_token in the sense of JWT validation?

**Comment by PinpointTownes** [comment]
> Just for understanding: JWT validation requires you to pass a resource param.

If you don't disable audience validation, yes. The OAuth2 bearer middleware is extremely strict by default: if your access tokens don't have an `aud` claim corresponding to the mandatory `options.Audience` option, it will return an exception. Yet, if you don't provide the `resource` parameter, your authorization server has no way to determine the audience, and it returns a token without any audience.

> OpenId Connect allows you to use JWT but does not enforce resource param to be able to create valid access_token in the sense of JWT validation?

OIDC specifications say nothing about the access token format, so yeah, officially, `resource` is not necessary to retrieve an access token.

If you don't want to use the `resource` parameter nor disable audience validation, you can also probably use the `AuthorizationEndpoint` and `TokenEndpoint` notifications to extract the `resource` parameter from another source (like `scope`).

**Comment by AlexZeitler** [comment]
Thanks.

Using scope is one thing I've already thought about but I did not come to a final conclusion until now 🙈

**Comment by AlexZeitler** [comment]
Did not forget about the sample and PR. Just lots of travelling :sob:

**Comment by PinpointTownes** [comment]
No problem :smile: 

I played with your sample, and it's pretty cool! I also tried finding a solution for the `>> Error: Cannot find module 'load-grunt-tasks'` bug, and it sounds like adding a `Gruntfile.js` file fixes it. 

(you'll need to fix a few conflicts when rebasing your PR, but nothing horrible :smile:)

**Comment by PinpointTownes** [comment]
FYI, @ilmax @Bartmax and I are currently working on a turnkey server for vNext named `OpenIddict`, that natively integrates ASP.NET Identity 3 and that comes with a default controller and default views (like IdentityServer3 :smile:). It's still very young, but if you want to try it or participate, don't hesitate!

https://github.com/openiddict/core
